src/docs/arch-concurrency.dox

```diff
@@ -0,0 +1,107 @@
+stores and code reordering. WiredTiger defines the following macros WT_READ_ONCE and
+WT_WRITE_ONCE, these macros, if used correctly, prevent the described forms of compiler
+optimization. Additionally some variables in WiredTiger are defined with the volatile keyword if
+it is known that they will be frequently accessed in a concurrent context.
+
+@section locking Locking
+
+There are three types of locks in WiredTiger:
+ - Condition Variables
+ - Spin Locks
+ - Read Write (RW) Locks
+Each mechanism guarantees that once a thread relinquishes its lock and another thread obtains the
+same lock, all modifications made by the first thread will be visible to the thread now holding
+the lock.
+
+There are two implementations of condition variables in WiredTiger respectively for POSIX and MSVC
+under the same interface. The specific implementation is determined at compile time based on the
+platform.
+
+There are three different implementations for spin locks in WiredTiger under the same interface:
+ - An implementation using GCC's builtin atomic operations.
+ - An implementation using a POSIX based mutex.
+ - An implementation for MSVC.
+Again specific implementation is determined at compile time based on the platform and compile-time
+configurations. WiredTiger has its own read-write lock implementation. It is based on atomic
+operations and condition variables.
+
+@subsection lock-usage Lock usage
+A correctly used lock takes the lock for each access to the shared memory it guards. WiredTiger
+sometimes deviates from this model and reads variables outside the context of the relevant lock.
+Such instances are only permitted if reading inconsistent data is sufficient for the use case.
+
+@section generations Generations
+
+Shared object, including the page index, WT_REFs, etc. have their lifecycles managed using
+the WiredTiger generations system. In general when a reader thread accesses an object, that could be
+concurrently freed, it enters a "generation". After entering that generation it accesses the object.
+A generation is a monotonically increasing integer which denotes times at which a given object must
+exist.
+
+Writer threads are then able to replace that object but cannot free it until the reader thread has
+left or released the generation. Instead the writer thread adds the object to a "stash" and it marks
+the stashed object with the generation indicating the point at which it was added to the stash. Once
+all reader in the system do not have a corresponding generation which is older than or equal to the
+stashed object's generation it can be freed.
+
+@section atomics Atomics
+
+As previously mentioned WiredTiger already expects loads and stores to specific sizes of points
+to be atomic and prevent tearing. WiredTiger also relies on atomic RMW operations, these include:
+- Compare and Swap
+- Fetch add / sub
+- Add / sub fetch
+
+Currently WiredTiger is undergoing a transition where all shared memory variable access will be
```

Comment:
I think the documentation should concern itself with the code "as it is" at the time this document is added/updated.

Reply:
To be fair the "as it is" state of the code is that WiredTiger is in the middle of transitioning to TSan, which matches the written documentation. That said I understand your point to be that documentation shouldn't really describe a transitional state.

This means I can either speak to pre or post TSan. I will opt to speak to the pre TSan form of the codebase which means to delete this paragraph in it's entirety. Is that acceptable @marcbutler?

Reply:
Is this expected to be a sustained transitional state: or is there a complete timeline for the TSan work? If there is a completion date: does it make sense to postpone updating this documentation until that work has been complete?

Reply:
I think we can just leave it out and create a ticket to add this paragraph at the completion of that work. Thoughts?

Reply:
My global observation is that we should defer documenting anything that is considered short-term unstable/transitional.